bin/docserv-build-navigation

```diff
@@ -119,23 +122,24 @@ for file in $template_html $template_translation $template_resources; do
   ([[ ! -d "$file" ]] && [[ ! -f "$file" ]]) && out "File/directory $file does not exist."
 done
 
-[[ ! $(grep -P '^'"$magic_comment"'$' $template_html) ]] && \
-  out "Template $template_html does not include magic comment: $magic_comment"
-
-ui_translation=$(comm -1 -2 <(echo "$doc_language" | tr ' ' '\n' | sort -u | head -1) <(echo -e "$ui_languages" | tr ' ' '\n' | sort -u))
-if [[ ! "$ui_translation" ]]; then
-  ui_translation=$default_ui_language
-fi
-
+# CREATE DIRECTORY TREE FOR OUTPUT
 temp_dir=$(mktemp -d /tmp/docserv-build-navigation-XXXXXXXX)
-intermediate_output=$temp_dir/page.html
-cache_file=$temp_dir/cache.xml
 
-output=$output_dir/index.${doc_language}.html
-[[ "$doc_language" == "$default_ui_language" ]] && output=$output_dir/index.html
-
-cache_files=$(find "$cache_dir/$doc_language" -name '*.xml')
+# Where to place the JSON data files
+data_path='docserv/data'
+# Where to place the template's resource files (JS, CSS, images)
+res_path='docserv/res'
+for product in $allproducts; do
+  mkdir -p $output_dir/$data_path/$product
+done
+for lang in "$ui_languages"; do
+  for product in $allproducts; do
+    mkdir -p $output_dir/$lang/$product
```

Comment:
The two for loops in line 132 and 135 could be merged like this:

```bash
for lang in "$ui_languages"; do
  for product in $allproducts; do
    mkdir -p $output_dir/$lang/$product
    mkdir -p $output_dir/$data_path/$product
done
```

Reply:
Oh wait, maybe this is better:

```bash
for product in $allproducts; do
   mkdir -p $output_dir/$data_path/$product
   for lang in "$ui_languages"; do
      mkdir -p $output_dir/$lang/$product
   done
done
```

I haven't tried it, it's just a small optimization idea. Use it or leave it. :wink:

Reply:
Not quite, for `$data_path`, the `$ui_language` is irrelevant.

Reply:
Yes, I know, but it doesn't matter. :-) 
You have two for loops which iterate two times over `$allproducts`. My suggestion takes these two separate for loops together so you don't need to repeat yourself. Saves you some CPU cycles. ;-)